### PR TITLE
Bump CAF to version 0.18.6

### PR DESCRIPTION
**Pull request by patszt**
This PR bumps the CAF version to 0.18.6.
Some of the major weird things happening in the review:

- Many config parsing now have .data() added to string_view. This is added because the get_or calls require the last argument (which is now std::string_view) to have a type_id. The type_id requires the type to be serializable. I dind't come up with a good way to deserialize into std::string_view as it doesn't own a buffer. This is why for now i have just used .data(). In CAF 0.19 this is fixed i believe

### CODE REVIEW

File: libvast/include/vast/qualified_record_field.hpp
Comment (patszt) [inline libvast/include/vast/qualified_record_field.hpp:85]: Changes done to the lack of inspect for std::string_view.

File: libvast/include/vast/system/spawn_arguments.hpp
Comment (patszt) [inline libvast/include/vast/system/spawn_arguments.hpp:53]: All of these are const references. The inspector api requires them to be non const so the easy way is just to create copies locally. Seems to be only used for serialization

File: libvast/src/system/exporter.cpp
Comment (patszt) [inline libvast/src/system/exporter.cpp:78]: One of the receivers depends on knowin who the sender is. In CAF 0.17.6 the anon_send didn't work properly and the sender was available. In CAF 0.18.6 the sender is nullptr so the code stopped working.

File: libvast_test/vast/test/test.hpp
Comment (patszt) [inline libvast_test/vast/test/test.hpp:141]: CAF_CHECK_FAIL was removed and the 0.17.6 implementation is quite weird. Didn't seem to be worth using anyways

File: libvast/test/table_slice.cpp
Comment (patszt) [inline libvast/test/table_slice.cpp:285]: Cannot use CHECK_NOT_EQUAL as these require non const operands for new inspector API. begin and end function return const ones

File: libvast/include/vast/detail/legacy_deserialize.hpp
Comment (patszt) [inline libvast/include/vast/detail/legacy_deserialize.hpp:34]: This is a copy paste code from CAF. I believe it requires some annotation

File: libvast/CMakeLists.txt
Comment (dominiklohmann) [inline libvast/CMakeLists.txt:178]: This must also be adapted in the docs (there's a table listing versions in the build from source section), and it also should have its own changelog entry that lists all the user-facing changes, e.g.,

> Building VAST now requires CAF 0.18.6. VAST supports setting advanced options for CAF directly in its configuration file under the `caf` section. If you were using any of these, compare them against the bundled `vast.yaml.example` file to see if you need to make any changes. The change has (mostly positive) [performance and stability implications](https://www.actor-framework.org/blog/2021-01/benchmarking-0.18/) throughout VAST, especially in high-load scenarios.

Similarly, the `vast.yaml.example` file should be updated to contain all the new CAF options rather than the old ones.

File: libvast/CMakeLists.txt
Comment (dominiklohmann) [inline libvast/CMakeLists.txt:224]: Note that you will have a merge conflict here once #2719 gets merged, as that removes the `VAST_ENABLE_OPENSSL` option and makes that mandatory.

File: plugins/broker/CMakeLists.txt
Comment (dominiklohmann) [inline plugins/broker/CMakeLists.txt:4]: Instead of this I propose that we remove the code outright and create a story for bringing it back. This also needs its own changelog entry. I propose the following:

> We removed the `broker` plugin that enabled direct Zeek 3.x log transfer to VAST. The plugin will return in the future written for Zeek 5+.

**Review by dominiklohmann — COMMENTED** [review]
First review round: Everything that relates to the build scaffolding and CI. I really like what I'm seeing so far!

File: libvast/include/vast/system/spawn_arguments.hpp
Comment (dominiklohmann) [reply libvast/include/vast/system/spawn_arguments.hpp:53]: This is a good approach I think. This is definitely not in the hot path and even then this would just be copying a few strings.

File: tools/lsvast/src/print_partition.cpp
Comment (dominiklohmann) [inline tools/lsvast/src/print_partition.cpp:75]: I propose the following for the error message:

> failed to deserialize value index: FlatBuffers table did not contain data field

File: tools/lsvast/src/print_segment.cpp
Comment (dominiklohmann) [inline tools/lsvast/src/print_segment.cpp:39]: We probably want to `exit(1)` here if printing failed.

I propose changing the error message like this:

```cpp
fmt::print(stderr, "!! failed to write segment slice: {}", err);
```

**Review by dominiklohmann — COMMENTED** [review]
Second part of the review: lsvast.

**Review by dominiklohmann — COMMENTED** [review]
Part three: All the builtins.

File: plugins/parquet/parquet.cpp
Comment (dominiklohmann) [inline plugins/parquet/parquet.cpp:37]: Should this use `detail::apply_all`?

File: plugins/web/tests/authentication.cpp
Comment (dominiklohmann) [inline plugins/web/tests/authentication.cpp:24]: Should this be a `CHECK_EQUAL(..., caf::error{})` instead so you can actually see what error it was in case one is returned?

**Review by dominiklohmann — COMMENTED** [review]
Part four: plugins.

File: libvast_test/src/events.cpp
Comment (dominiklohmann) [inline libvast_test/src/events.cpp:110]: This should compare against `caf::error{}` so the error is shown if there is one. It should also be `REQUIRE_*` rather than `CHECK_*` as it doesn't make sense to continue test execution if this fails.

File: libvast_test/vast/test/test.hpp
Comment (dominiklohmann) [inline libvast_test/vast/test/test.hpp:30]: Is this still needed?

File: libvast_test/vast/test/test.hpp
Comment (dominiklohmann) [inline libvast_test/vast/test/test.hpp:47]: This looks _really_ dangerous, even just for unit tests. What do we need this for?

File: libvast_test/vast/test/test.hpp
Comment (dominiklohmann) [reply libvast_test/vast/test/test.hpp:141]: I don't think I've ever used this, so that makes sense.

File: libvast_test/vast/test/type_ids.hpp
Comment (dominiklohmann) [inline libvast_test/vast/test/type_ids.hpp:31]: I don't think you need this at all if you wrap the `view<{list,map,record}>` objects into a `data_view` with `make_data_view(...)`.

Generally, view's should not be inspectable as they cannot be deserialized into by design.

File: libvast_test/vast/test/test.hpp
Comment (dominiklohmann) [reply libvast_test/vast/test/test.hpp:141]: I don't think I've ever used this, so that makes sense.

File: libvast/fbs/partition.fbs
Comment (dominiklohmann) [inline libvast/fbs/partition.fbs:45]: Please add a comment that we can mark this with a `(deprecated)` attribute once the minimum partition version is 3. The current version is 2, so for partitions that have version 3 we'll know that they cannot have this field set, so once we bumped the minimum to 3 we can remove the code generation for this field.

The same applies to the `combined_layout` field.

File: libvast/fbs/partition.fbs
Comment (dominiklohmann) [inline libvast/fbs/partition.fbs:62]:
```suggestion
  /// The schema of a partition.
  schema: [ubyte] (nested_flatbuffer: "vast.fbs.Type");
```

File: libvast/fbs/synopsis.fbs
Comment (dominiklohmann) [inline libvast/fbs/synopsis.fbs:9]:
```suggestion
  /// NOTE: This can be marked (deprecated) once the minimum partition version is 3.
  caf_0_17_data: [ubyte];
  
  /// A synopsis serialized using the CAF 0.18 binary serializer.
  caf_0_18_data: [ubyte];
```

File: libvast/fbs/value_index.fbs
Comment (dominiklohmann) [inline libvast/fbs/value_index.fbs:24]: Please also add comments along the lines of `/// NOTE: This can be marked (deprecated) once the minimum partition version is 3.` to the two older fields, and please copy over the comments from the two old fields to the new ones. They get copied over by the FlatBuffers compiler into the generated code, so they show up in auto-completion and can be really helpful.

File: libvast/fbs/value_index.fbs
Comment (dominiklohmann) [inline libvast/fbs/value_index.fbs:20]: Style nit: All the other adjusted fields had `caf_0_x` as a prefix, this and its replacement have it as a suffix.

**Review by dominiklohmann — COMMENTED** [review]
Part six: FlatBuffers tables.

File: libvast/include/vast/detail/settings.hpp
Comment (patszt) [inline libvast/include/vast/detail/settings.hpp:43]: I am not really happy with the use of this function.
What i tried to achieve is to create a wrapper for caf::config_option_set that knows how to handle the list values in the same manner as this function. I would replace the usage of caf::config_option_set with another class and no one would need to use this free function again.
By doing .add<std::vector> and then .parse() it would all be handled correctly.

With current approach it is needed to call this free function whenever we use list config values
Unfortunately the proposed change breaks the plugin API interface so that would require additional work and woould be a breaking change for users that wrote their own plugins. I think it is worth having this as a separate story.

File: libvast/include/vast/concept/convertible/data.hpp
Comment (dominiklohmann) [inline libvast/include/vast/concept/convertible/data.hpp:101]: I don't think this is a user-facing error. Can we make this an assertion for this function instead and then just assume below?

File: libvast/include/vast/fbs/utils.hpp
Comment (dominiklohmann) [inline libvast/include/vast/fbs/utils.hpp:65]:
```suggestion
    return caf::make_error(ec::serialization_error, "failed to serialize bytes");
```

File: libvast/include/vast/flatbuffer.hpp
Comment (dominiklohmann) [inline libvast/include/vast/flatbuffer.hpp:223]:
```suggestion

  consteval static auto qualified_name() noexcept -> std::string_view {
```

File: libvast/include/vast/http_api.hpp
Comment (dominiklohmann) [inline libvast/include/vast/http_api.hpp:85]:
```suggestion
```

File: libvast/include/vast/system/node_control.hpp
Comment (dominiklohmann) [inline libvast/include/vast/system/node_control.hpp:49]:
```suggestion
    auto str = std::string{in.data(), in.data() + in.size()};
```

**Review by dominiklohmann — COMMENTED** [review]
Review round seven: About the first half of the libvast; I'm done for the day.

File: libvast/include/vast/concept/convertible/data.hpp
Comment (patszt) [reply libvast/include/vast/concept/convertible/data.hpp:95]: I would revisit this comment for CAF 0.19 bump. Maybe in this version it will be doable.

File: .github/workflows/vast.yaml
Comment (patszt) [reply .github/workflows/vast.yaml:114]: It's also used for build vast jobs. It was moved from optional cmake-extra-flags into here as it was common to both debian and macOS.

File: .github/workflows/vast.yaml
Comment (patszt) [reply .github/workflows/vast.yaml:610]: brew installed CAF is 0.18.5. I have no idea if there is a workaround to install 0.18.6. Maybe we can request some BUMP in brew?

File: .github/workflows/vast.yaml
Comment (dominiklohmann) [reply .github/workflows/vast.yaml:114]: It doesn't make sense for me here as this is a variable called `cmake_version_args`, and whether CAF is bundled or not should not be part of that.

File: .github/workflows/vast.yaml
Comment (dominiklohmann) [reply .github/workflows/vast.yaml:610]: Can we build with 0.18.5+? Then we could also speed up the Debian CI and Docker image build by using the one that's available upstream.

If not then I can request a bump on Homebrew, but this would then be in a follow-up.

File: tools/lsvast/src/print_partition.cpp
Comment (patszt) [reply tools/lsvast/src/print_partition.cpp:63]: I think pointing it out in this context is not necessarily helpful.

File: libvast_test/vast/test/type_ids.hpp
Comment (patszt) [reply libvast_test/vast/test/type_ids.hpp:31]: I think this is triggered in CHECK_EQUAL(data_view, data_view)

data_view is a caf::variant and it requires all types to be inspectable in inspect(auto& inspector, caf::variant);

I could replace the mentioned code by replacing comparisons with data_view into materialize(data_view)
It should work, right?

Comment (dominiklohmann) [reply libvast_test/vast/test/type_ids.hpp:31]: Yes, that would work. It's probably safer to do that way.

File: libvast_test/vast/test/test.hpp
Comment (patszt) [reply libvast_test/vast/test/test.hpp:47]: It is related to data_view comparison i believe. I should be able to remove it

File: plugins/broker/CMakeLists.txt
Comment (patszt) [reply plugins/broker/CMakeLists.txt:4]: I will make a separate PR for this change

Comment (patszt) [reply plugins/broker/CMakeLists.txt:4]: Before i merge this one i mean

File: .github/workflows/vast.yaml
Comment (patszt) [reply .github/workflows/vast.yaml:610]: We can for macOS. For debian we can download only 0.17.6 though

Comment (dominiklohmann) [reply .github/workflows/vast.yaml:610]: Check. Makes sense.

File: libvast/include/vast/detail/legacy_deserialize.hpp
Comment (dominiklohmann) [reply libvast/include/vast/detail/legacy_deserialize.hpp:34]: I think this is fine as-is if you move it to right above where you use it and then `#undef` it again immediately afterwards.

File: libvast/include/vast/detail/settings.hpp
Comment (dominiklohmann) [reply libvast/include/vast/detail/settings.hpp:43]: If I understood correctly then you already created this story, so we can probably resolve this. For now I think this is fine.

File: libvast/src/format/json.cpp
Comment (dominiklohmann) [inline libvast/src/format/json.cpp:1044]: We shouldn't be able to run into this error, as that would be a precondition violation. Please check `!err` with an assertion.

File: libvast/src/format/json.cpp
Comment (dominiklohmann) [inline libvast/src/format/json.cpp:1057]: We shouldn't be able to run into this error, as that would be a precondition violation. Please check `!err` with an assertion.

File: libvast/src/format/test.cpp
Comment (dominiklohmann) [inline libvast/src/format/test.cpp:358]: If finish fails here we should return the error from the function.

File: libvast/src/segment.cpp
Comment (dominiklohmann) [inline libvast/src/segment.cpp:238]: If this fails we should return the error from the function.

File: libvast/src/store.cpp
Comment (dominiklohmann) [inline libvast/src/store.cpp:405]: What's the context of this change?

Please also rewrite the log message in the usual style, e.g.,

> store encountered unexpected end of input before shutdown

File: libvast/src/system/default_configuration.cpp
Comment (dominiklohmann) [inline libvast/src/system/default_configuration.cpp:52]: Can you write a story for re-evaluating the CAF defaults we set with VAST after merging this PR? I think we should definitely do this before cutting the next release.

File: libvast/src/system/disk_monitor.cpp
Comment (dominiklohmann) [inline libvast/src/system/disk_monitor.cpp:153]: Is that static cast still necessary? I thought that'd be fixed :(

**Review by dominiklohmann — COMMENTED** [review]
Round eight: The remainder of libvast.

File: libvast/src/system/disk_monitor.cpp
Comment (patszt) [reply libvast/src/system/disk_monitor.cpp:153]: It is :/

File: libvast/src/store.cpp
Comment (patszt) [reply libvast/src/store.cpp:405]: This is the issue with our source streams propagating ec::end_of_input error when the input ended.
Previously this error couldn't propagate here for some reason. Now it is sometimes visible and sometimes not. I believe the error was only introduced to signalize the end of input so we kind of ignore it instead of treating it like a real error

Comment (dominiklohmann) [reply libvast/src/store.cpp:405]: Okay, that makes sense. We need to re-evaluate the stream on import with pipelines anyways, so I think we can solve this sometime in the future rather than now. Please just rephrase the error message then.

File: libvast/test/system/configuration.cpp
Comment (dominiklohmann) [inline libvast/test/system/configuration.cpp:65]:
```suggestion
      FAIL("failed to unpack " << name << " to vector");
```

**Review by dominiklohmann — APPROVED** [review]
I mostly glanced at the tests as the changes were mostly mechanical. Initial testing shows that everything is working, even with an old database.

I would like to see this merged and then roll this out on our testbed for performance testing and to see what we need to tweak with the defaults.

There's one last warning I get when building, and this diff fixes it:

```diff
diff --git a/libvast/test/format/arrow.cpp b/libvast/test/format/arrow.cpp
index 54f3bacd9..a92fa5a0b 100644
--- a/libvast/test/format/arrow.cpp
+++ b/libvast/test/format/arrow.cpp
@@ -94,7 +94,8 @@ TEST(arrow IPC read) {
   format::arrow::writer writer;
   writer.out(stream);
   for (auto& slice : zeek_conn_log)
-    writer.write(slice);
+    if (auto err = writer.write(slice))
+      FAIL("failed to write zeek.conn log " << err);
   auto data = stream->Finish().ValueOrDie()->ToString();
   auto in = std::make_unique<std::istringstream>(std::string{data});
   auto options = caf::settings{};
```

This PR is a fantastic piece of work 🚀

File: libvast/CMakeLists.txt
Comment (dominiklohmann) [reply libvast/CMakeLists.txt]: I've written a follow-up story for the vast.yaml.example file so we can unblock this.

File: libvast/src/system/default_configuration.cpp
Comment (dominiklohmann) [reply libvast/src/system/default_configuration.cpp:52]: I have written this follow-up story.

**Review by dominiklohmann — APPROVED** [review]
I just pushed a bunch of things—only action left to do (assuming that CI runs through) is to squash and then merge!

**Pull request merged**
